src/AspNet.Security.OAuth.Salesforce/SalesforceAuthenticationHandler.cs

```diff
@@ -0,0 +1,83 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (http://www.apache.org/licenses/LICENSE-2.0)
+ * See https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers
+ * for more information concerning the license and the contributors participating to this project.
+ */
+
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Security.Claims;
+using System.Threading.Tasks;
+using AspNet.Security.OAuth.Extensions;
+using JetBrains.Annotations;
+using Microsoft.AspNetCore.Authentication;
+using Microsoft.AspNetCore.Authentication.OAuth;
+using Microsoft.AspNetCore.Http.Authentication;
+using Microsoft.AspNetCore.WebUtilities;
+using Microsoft.Extensions.Logging;
+using Newtonsoft.Json.Linq;
+
+namespace AspNet.Security.OAuth.Salesforce
+{
+    public class SalesforceAuthenticationHandler : OAuthHandler<SalesforceAuthenticationOptions>
+    {
+        public SalesforceAuthenticationHandler([NotNull] HttpClient client)
+            : base(client)
+        {
+        }
+
+        protected override async Task<AuthenticationTicket> CreateTicketAsync([NotNull] ClaimsIdentity identity,
+            [NotNull] AuthenticationProperties properties, [NotNull] OAuthTokenResponse tokens)
+        {
+            var identityServiceUrl = tokens.Response.Value<string>("id");
+            var salesforceInstanceUrl = tokens.Response.Value<string>("instance_url");
+
+            var address = QueryHelpers.AddQueryString(identityServiceUrl, "token", tokens.AccessToken);
+            var request = new HttpRequestMessage(HttpMethod.Get, address);
+            request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", tokens.AccessToken);
+            request.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
+
+            var response = await Backchannel.SendAsync(request, HttpCompletionOption.ResponseHeadersRead, Context.RequestAborted);
+            if (!response.IsSuccessStatusCode)
+            {
+                Logger.LogError("An error occurred when retrieving the user profile: the remote server " +
+                                "returned a {Status} response with the following payload: {Headers} {Body}.",
+                                /* Status: */ response.StatusCode,
+                                /* Headers: */ response.Headers.ToString(),
+                                /* Body: */ await response.Content.ReadAsStringAsync());
+
+                throw new HttpRequestException("An error occurred when retrieving the user from the Salesforce identity service.");
+            }
+
+            var payload = JObject.Parse(await response.Content.ReadAsStringAsync());
+            Logger.LogTrace(payload.ToString());
+            identity.AddOptionalClaim(ClaimTypes.NameIdentifier, SalesforceAuthenticationHelper.GetUserIdentifier(payload), Options.ClaimsIssuer)
```

Comment:
Omagad, this list is insane :smile: 
Adding all these claims will result in extremely large cookies, that may excess the browsers size limits.

Please consider reducing the list to 5/6 claims.